src/viperleed/calc/lib/dataclass_utils.py

```diff
@@ -0,0 +1,227 @@
+"""Module dataclass_utils of viperleed.calc.lib.
+
+Collects useful functions that add up on top of the dataclasses
+stdlib module.
+"""
+
+__authors__ = (
+    'Michele Riva (@michele-riva)',
+    )
+__copyright__ = 'Copyright (c) 2019-2024 ViPErLEED developers'
+__created__ = '2024-07-25'
+__license__ = 'GPLv3+'
+
+import copy
+import functools
+from dataclasses import FrozenInstanceError
+from dataclasses import dataclass
+from dataclasses import field as data_field
+from dataclasses import fields as data_fields
+from dataclasses import is_dataclass
+from dataclasses import replace as data_replace
+import typing
+import sys
+
+try:  # Import stuff that was made available since py3.8
+    from typing import get_args as ty_get_args
+except ImportError:
+    from typing_extensions import get_args as ty_get_args
+    from typing_extensions import get_origin as ty_get_origin
+else:
+    from typing import get_origin as ty_get_origin
+
+if sys.version_info < (3, 12):  # No frozen_default keyword in py3.11
+    from typing_extensions import dataclass_transform
+else:
+    from typing import dataclass_transform
+
+
+frozen = dataclass_transform(frozen_default=True)(
+    functools.partial(dataclass, frozen=True)
+    )
+
+
+def check_types(self, init_only=False):
```

Comment:
Do we need to implement this ourselves? Doesn't pylint do this anyways if there are type annotations?

Reply:
Unfortunately type checking for `dataclasses` in `pylint` is still pretty rudimentary. Works OK for explicit types, but not for fields that use the `dataclasses.field` function. (See the plenty issues concerning `dataclasses` at https://github.com/pylint-dev/pylint/issues